rfc/build-system/build-system.md

```diff
@@ -0,0 +1,711 @@
+that dependency are generated by these builds. This system is also used by the
+plugin system for versioning the plugin APIs.
+
+The other system is used for the artifacts we distribute to the community:
+Docker containers and Debian packages. This system runs in Jenkins and is a
+complicated mash-up of Jenkins and Google Cloud Build configuration, shell
+scripts, Python, and Kotlin. Some of it is checked-in and some of it just lives
+in Jenkins configs or files on the Jenkins server.
+
+These build systems are mostly unaware of each other. They use different version
+numbering schemes, but with identically-shaped version numbers. For example, in
+the past year, they've each generated an echo 2.12.0, but those two 2.12.0s
+aren't related to one another in any way. Their git tags are slightly different
+(`v2.12.0` vs. `version-2.12.0`) and their artifacts are published to different
+Bintray repositories, so there are no conflicts. (Yes, we also have two
+different Bintray repositories.)
+
+## Timeline
+
+This build system will be implemented between the 1.22 release in mid-August and
+the following release (which will no longer be called 1.23) in mid-October.
+
+## Design
+
+### New Versioning Scheme and Release Schedule
+
+Spinnaker releases will switch to date-based versions, so instead of 1.23, we
+will call it 2020.4. The release schedule will change from every 8 weeks to
+every 3 months.
+
+Anecdotally, we believe that some users expect that there won't be any breaking
+changes between 1.20 and 1.21, perhaps confusing Spinnaker versions with
+[SemVer](https://semver.org/). In fact, we regularly introduce breaking changes
+in these releases.
+
+Additionally, this new version scheme will help users know how old their
+Spinnaker version is; we know many users are running very old versions of
+Spinnaker and this may goad them into upgrading more frequently.
+
+A third, lesser advantage of date-based versions is that it makes a clear break
+from microservice versions of the past. For example, there is already an Orca
+1.22.0 in one of the old version schemes.
+
+The release branches will be cut on the second Wednesday of the quarter, and the
+release will be targeted for the following Wednesday. (These dates were chosen
+to avoid major US holidays.) The time between releases will now be either 13 or
+14 weeks, depending on the calendar. The next releases will be:
+
+- 2020.4: branches cut 2020-10-14, release on 2020-10-21
+- 2021.1: branches cut 2021-01-13, release on 2021-01-20
+- 2021.2: branches cut 2021-04-14, release on 2021-04-21
+- 2021.3: branches cut 2021-07-14, release on 2021-07-21
+
+Slowing the release cadence reduces the release manager overhead somewhat,
+although release manager shifts are now longer and may have to be shared by
+multiple people. It also helps many of our users who find upgrading to be a lot
+of work by reducing how frequently they have to upgrade. It does mean it takes
```

Comment:
Do we have any data about why users feel like upgrading is a lot of work?

Reply:
I don't even know what the upgrade process looks like for a typical user looking for zero downtime.

I know at Netflix we try to be aware of the order in which we're upgrading services and even our best attempts are occasionally foiled.

Reply:
I'd say that most, if not all, OSS users just run `hal config version edit --version {target-version}` and deploy. Kubernetes does a decent enough job of rotating things without downtime. The majority of the time, an unknown, breaking change doesn't present until things have been successfully deployed.

Reply:
Gotcha. 

Surprise that doesn’t break often (or maybe it does). We still occasionally get tripped up by coupled changes across services. Changes that in isolation don’t look like traditional breaking changes.

Reply:
It's not the upgrade itself that is a lot of work. It's reading or trying to read all the release notes and commits to prepare for upgrading and then the verification that the upgrade didn't regress that is a lot of work.